### ISSUE TITLE
Issues with creating cluster matrix

### ISSUE THREAD

**Issue by bentsherman** [SystemsGenetics/KINC#24]
I'm still having issues reading the cluster matrix produced by GMM. Again, I'm not sure if it's a problem with my code or the `CCMatrix` code since there aren't any other analytics that produce this data type. As a quick experiment, I add this code to `GMM::runSerial()` in order to produce a cluster matrix with just the first few pairs in a few seconds:
```
   // increment through all gene pairs
   int numPairs = 1000;
   while ( numPairs-- > 0 && vector.geneX() < _output->geneSize() )

   // also remove OpenCL capability in gmm.h to force runSerial()
```

Then when I try to open the cluster matrix I get this error repeatedly:
```
Attempting to seek to gene pair 754 when total size is 607.
File: ../src/genepair_base.cpp
Function: void GenePair::Base::seekPair(qint64) const
Line: 277
```

As far as I can tell, inserting that 1000-pair limit shouldn't affect the validity of the cluster matrix, it only lowers the number of pairs that are written to it. But to be sure I ran GMM completely on Yeast and I encountered the same error but with larger numbers. Am I doing something wrong with creating the cluster matrix? I don't really understand why this is happening.

**Comment by 4ctrl-alt-del** [comment]
That sounds like a bug in CCMatrix to me, I will attempt to replicate with the code you provide and follow up as soon as I know more.

**Comment by 4ctrl-alt-del** [comment]
OK I think I have fixed this. It was indeed a bug in the GenePair::Base class used by cluster matrix. The reason this bug never came up for me is I have only used a single cluster for any analytic output I have done and this bug would only show itself if you have larger clusters.

Test it out and see if this fixes your issue. Bug fixed on commit 4f0133b363c8976ca58fb5610c3643aa1fe49033. I will let you close this issue if it is fixed for you.

**Comment by bentsherman** [comment]
Now I get this error:
```
Gene Pair Vector Error

Cluster 97 is outside limits.

1. File: ../src/genepair_vector.cpp
2. Function: qint64 GenePair::Vector::indent(qint8) const
3. Line: 37
```

I made sure to rebuild KINC with the update, and I deleted and recreated the CCM. Not sure where the 97 is coming from. Do you get this error?

**Comment by 4ctrl-alt-del** [comment]
I do not get that error, it works for me. The maximum number of clusters is 64... would you be making more than that?

**Comment by bentsherman** [comment]
No, I only go from 1 to 5. In fact, here are the commands that I use:
```
kinc run import_emx --input "data/Yeast-GEM.txt" --output "data/Yeast.emx" --nan "NA" --size 0 --transform "none"
kinc run gmm --input "data/Yeast.emx" --output "data/Yeast.ccm" --min 30 --minthresh -inf --minclus 1 --maxclus 5 --crit "BIC" --bsize 4 --ksize 4096
```

Could you send me your CCM? It should only be a few hundred KB, I can at least try to see if I can open yours. I'm wondering if the difference is due to using different Qt versions or environments.

Edit: Tried the same thing on Palmetto with Qt 5.9.2, got the same error. On my laptop I have Qt 5.6.2.

**Comment by 4ctrl-alt-del** [comment]
I was going to send the file but it is much bigger than KB... 352 MB in size. That is way too big to send by email.

**Comment by bentsherman** [comment]
Well that's a major discrepancy, my CCM is only about 500 kB. Not sure what would cause that.

**Comment by 4ctrl-alt-del** [comment]
My mistake, I realized why it was that big; I kept overwriting it from previous runs of GMM where I did the whole thing. It is less than a megabyte now you should get an E-Mail shortly. Like I said, it opens without error for me.

**Comment by bentsherman** [comment]
I was able to open your CCM. Didn't see any clusters, but there are probably just so few of them that they're hard to find. In any case it worked. Are you developing on Windows? If so could you try to test it on Linux? At this point I think that's probably the source of this error.

Also I will email you my CCM so that you can try to view it on either OS. Note in particular that it is slightly smaller, which suggests a problem in how the CCM is saved.

**Comment by bentsherman** [comment]
A few other random notes from today: (1) I can open a full CCM which I generated a few weeks ago, which is about 8 GB, and (2) now I get a different error when I try to open a minimal CCM:
```
Cannot initialize gene vector (1379088437,1459836330).

1. File: ../src/genepair_vector.cpp
2. Function: GenePair::Vector::Vector(qint32, qint32)
3. Line: 20
```

This seems to happen with the latest commit which adds outlier removal. Not sure how this commit changes the error because outlier removal is disabled by default. So I further suspect that there is some weird environment issue going on.

**Comment by 4ctrl-alt-del** [comment]
I have not touched windows anything in almost a decade. Not only do I run in Linux, I run in Gentoo Linux. :)

**Comment by 4ctrl-alt-del** [comment]
My gut is telling me file corruption from the errors you are getting. The errors you are getting comes from the "header" of each cluster getting junk values from the data file at the position in the file they should exist.  The header is (geneX 32b)(geneY 32b)(cluster 8b). All the errors you are getting is the genepair base class trying to create genepair vectors that are way way beyond it's limits which tells me it is getting junk data from the file.

I wish I could help more but this doesn't happen for me.

**Comment by bentsherman** [comment]
Yeah, this has been my growing suspicion also. I think at this point I know enough that I can scrutinize the gene pairs as they are written to the output file and hopefully find the source that way.

**Comment by bentsherman** [comment]
After exploring the `GenePair::Base` a bit, I tried printing some data as it is read from a CCM file:
```
// in Base::readData()
qInfo() << _geneSize << _dataSize << _pairSize << _rawPairSize << _offset;

// in Base::findPair()
qInfo() << geneX << geneY << cluster;
```

And when I open the CCM that you generated, I get this output:
```
19898 69 47 121 4
1437204480 737620 49
1437204480 737620 49
1437204480 737620 49
842346542 822302122 0
1096036679 808726580 48
959721518 822302122 0
1437204480 737620 49
...
```

And the output continues in that pattern. These gene pair vectors / clusters are definitely invalid, but your CCM doesn't throw the two errors I get because (1) geneX is always greater than geneY and (2) the cluster values are all below 64. If you were to assert that the gene vector be within `_geneSize` or that the cluster sizes be no greater than 5, you would get an error too.

More importantly, your CCM doesn't display any of the clusters, but there are definitely clusters in that top-left corner because they are from the first 1000 gene pairs. So now I think there is still a problem within `GenePair::Base` or `CCMatrix`. Otherwise you should be able to open your CCM and immediately see clusters in the first few rows.

**Comment by bentsherman** [comment]
Also on a slightly related note, I don't know how much this number was discussed, but in case the max cluster size influences the total file size of a CCM (which it looks like it does), I think this constant could be decreased significantly, to 32 or even 16.

**Comment by 4ctrl-alt-del** [comment]
Well that proves something is wrong with CCMatrix. I will start debugging at the lowest levels possible and let you know when I find something.

Reducing max cluster size would have zero effect on the file size. You in fact could increase it to 256 and still have no effect on the file size.

**Comment by 4ctrl-alt-del** [comment]
So after spending lots of time finding nothing wrong with the cluster matrix code I decided to make a dummy analytic ccmtest and created a cluster matrix with dummy data that is easy to validate. The dummy analytic creates a random amount of clusters per gene pair just to spice things up.

Guess what? It works flawlessly. The cluster matrix opens with zero error and presents the data that was given to it by the dummy analytic ccmtest. At this point I am 99% sure something in your gmm analytic is overwriting something it shouldn't or something similar.

Can you please verify the same results by checking out the branch I just made called ccmtest and running the dummy analytic yourself?

**Comment by bentsherman** [comment]
Hmm. Well, the CCM test runs smoothly, and when I open the cluster matrix, it works at first, but as I explore the matrix I begin to get that same error with initializing an invalid `GenePair::Vector`. Are you able to view the entire matrix without an error?

My use of CCMatrix is pretty much contained to `GMM::initialize()` and `GMM::savePair()`. If you see anything wrong let me know, because I don't. As far as I can tell I'm using CCMatrix the same way as your test code.

I will keep looking at this when I have time but I'm not really sure what else to check. I'll email you my output from CCM test so that you can compare and make sure you can open mine.

**Comment by 4ctrl-alt-del** [comment]
I can scroll through the entire cluster matrix window without a single error and just to be sure I looked at every single value string for every single gene pair and they are all correct. Not a single error.

However when I run your gmm it produces a corrupt cluster matrix file with massive amounts of errors.

At this point I don't know how I can help. I am very busy and behind on completing MPI so I need to focus on that first and finish it so I won't be able to do more with this until I come back from a small vacation middle of next week. I will look at GMM::initialize() and GMM::savePair() first thing next week when I get back from vacation.

**Comment by bentsherman** [comment]
Did you try the CCM I sent you? Because I couldn't open that one without errors either.

**Comment by bentsherman** [comment]
I am going to generate a full CCM to see if it is valid. At this point that is the only difference that I can see.

**Comment by bentsherman** [comment]
The full CCM that I generated seems to be valid. It was generated on Palmetto so I'm not sure if it's the fact that it was Palmetto or the fact that I generated the whole thing that made it work. I can use this CCM to test the correlation analytics in the mean-time.

**Comment by bentsherman** [comment]
I realized something that may help narrow our search for this bug: it seems to only happen when multiple clusters are written. This would explain why I never had issues with the correlation analytics even though they also use `GenePair::Base`, because without a CCM (or with an empty one), the output CMX will never have more than one cluster.

I confirmed this further with the CCM test: if I change the code to always write a single cluster, I can read the output CCM with no issue. However, if I allow the clusters to vary (as in the original code), I reach errors. So I think there must be something about writing multiple clusters that causes file corruption or something.

**Comment by 4ctrl-alt-del** [comment]
I thought the same myself. Which is why I made the CCM test analytic the way I did. Really wish it would break for me too so I could figure this thing out. After I am done redoing metadata and then fixing MPI I will be focusing on this issue until the culprit is found.

**Comment by bentsherman** [comment]
Sure, well when you get to this issue I think the natural next step is to try to match environments. My laptop environment is Qt 5.7 on Ubuntu 16.04, and my Palmetto environment is Qt 5.9 on some version of RHEL. Both environments seem to get the same errors, so if you can get access through Stephen to Palmetto (or Kamiak) or if you have an Ubuntu system, you could run the CCM test there.

**Comment by bentsherman** [comment]
I just noticed that the CCM that you sent me a while back is from a different GEM than what I've been using, which is the Yeast dataset. So I would also be sure to try that dataset too, using the trick to just compute the first 1000 clusters. That's what I was talking about before -- you should see a bunch of clusters in the top-left corner of the Yeast CCM if it saves the data correctly. Even if there's no error but it's completely blank, that means the data didn't save correctly.

**Comment by bentsherman** [comment]
Earlier I said that I was able to generate a full CCM on Palmetto that didn't have errors, but it turns out that this CCM is also corrupt, I guess the corruption doesn't occur until later in the file, so I didn't encounter it while viewing the matrix. So at this point I can say that the corruption has occurred consistently for me with the Yeast dataset.

It almost seems as if the corruption doesn't occur immediately, but towards the middle or the end. That is why, if you only compute a few gene pairs, opening the matrix immediately causes an error, whereas if you compute many more gene pairs, the farther you can scroll before encountering corrupted data.

**Comment by spficklin** [comment]
This may be a really naive question.... comes from me not being able to keep up with ACE/KINC as fast as you guys are developing.  Is there only one process that writes to the output file?  What I'm wondering is if there is any type of race conditions from two processes trying to write to the file at the same time and the longer it runs the more likely it is that this could occur?

**Comment by bentsherman** [comment]
There's only one process writing to the file, even in MPI only the master node saves the data. Running for longer doesn't seem to increase the probably of an error, rather it seems to happen every time, just towards the middle or end of the data rather than the beginning or everywhere.

**Comment by spficklin** [comment]
Does the size of the input GEM make a difference?  If you have fewer genes or more genes?

**Comment by bentsherman** [comment]
It's possible. I've only tried the Yeast dataset (7050 genes) which gives errors, but I think Josh has only tried some other dataset which I can't identify because it was just called "GEM.txt" but it has about 20,000 genes (first gene is AT1G09090.1 if that helps). That's another difference that could be a factor.

**Comment by spficklin** [comment]
If it doesn't bloat the repository, perhaps we could create an examples directory so that we all have the same samples for testing, and when we get around to implement unit testing for KINC we'll have those ready to go.   I'll try to test as well, although I can't promise as things just keep coming and I can't find a moment to do this fun stuff!

**Comment by bentsherman** [comment]
Yeah I think it would be good to have that for unit tests in the future. The KINC repo is actually pretty boated right now... almost 200 MB, probably because there were once large files in the repo that were deleted. We might want to clean the git history before we start adding data files.

Also, if you or @4ctrl-alt-del could send me the 20,000 gene data set that he was using, I could test that one on my system.

**Comment by bentsherman** [comment]
I fixed the bug! I frickin' did it.

In `GenePair::Base` there was a misuse of `_pairSize`, the total number of gene pairs, and `_rawPairSize`, the total number of gene pair clusters. The `getDataEnd()` method was using the former where it should have used the latter, which I think meant that metadata would be written right smack in the middle of the file somewhere instead of at the end. This error only occurs when there are multiple clusters, because otherwise the two variables are equal. I didn't see this before because I didn't understand the difference between these two variables, so I renamed `_rawPairSize` to `_clusterSize` for clarity.

As for why you couldn't replicate the bug, I think it was mostly because you were using a different dataset which I didn't realize at first. The only thing I can't explain is why the CCM test worked for you but not for me... maybe you just got lucky!

After months and much discussion... closing this issue.

**Comment by feltus** [comment]
Woohoo!

**Comment by 4ctrl-alt-del** [comment]
Great job Ben. I am even more perplexed that bug did not show up in the tests I did. With the multiple cluster test analytic I did I should have gotten very corrupt data... yet I did not. That's very strange.

The funny thing is the other bug I found was in one of the write routines using pairSize instead of rawPairSize and fixing that fixed it for me... but it should not have with the other mismatch.
